futures-util/src/stream/recover.rs

```diff
@@ -0,0 +1,35 @@
+use std::marker::PhantomData;
+
+use futures_core::{Stream, Poll, Async};
+use futures_core::task;
+
+/// Future for the `recover` combinator, handling errors by converting them into
+/// an `Option<Item>`, compatible with any error type of the caller's choosing.
+#[must_use = "futures do nothing unless polled"]
+#[derive(Debug)]
+pub struct Recover<A, F, E> {
+    inner: A,
+    f: Option<F>,
```

Comment:
You can just use `F` here (rather than `Option<F>`, since the closure is `FnMut`, and we want to be able to `recover` from multiple errors.